libmamba/src/core/history.cpp

```diff
@@ -325,4 +325,183 @@ namespace mamba
             out << specs_output("neutered", entry.neutered);
         }
     }
+
+    namespace detail
+    {
+        // The following function parses the different formats that can be found in the history
+        // file. conda/mamba1 format: installed: +conda-forge/linux-64::xtl-0.8.0-h84d6215_0
+        // removed: -conda-forge/linux-64::xtl-0.8.0-h84d6215_0 mamba2 broken format: installed:
+        // +conda-forge::xtl-0.8.0-h84d6215_0 removed:
+        // -https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0 mamba2 new format:
+        // installed: +https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0
+        // removed: -https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0
+        specs::PackageInfo pkg_info_builder(std::string s)
+        {
+            std::string s_pkg;
+            std::string channel;
+            std::size_t pos_0 = s.rfind("/");
+            if (pos_0 != std::string::npos)
+            {
+                std::string s_begin = s.substr(0, pos_0);
+                std::string s_end = s.substr(pos_0 + 1, s.size());
+                std::size_t pos = s_begin.rfind("/");
+                if (pos != std::string::npos)
+                {
+                    channel = s_begin.substr(pos + 1, s_begin.size());
+                }
+                else
+                {
+                    channel = s_begin;
+                }
+                s_pkg = std::get<1>(util::rsplit_once(s_end, "::"));
+            }
+            else
+            {
+                channel = std::get<0>(util::split_once(s, "::"));
+                s_pkg = std::get<1>(util::rsplit_once(s, "::"));
+            }
+
+            std::size_t pos_1 = s_pkg.rfind("-");
+            std::string s_pkg_ = s_pkg.substr(0, pos_1);
+            std::string build_string = s_pkg.substr(pos_1 + 1, s_pkg.size());
+
+            std::size_t pos_2 = s_pkg_.rfind("-");
+            std::string name = s_pkg_.substr(0, pos_2);
+            std::string version = s_pkg_.substr(pos_2 + 1, s_pkg_.size());
+
+            specs::PackageInfo pkg_info{ name, version, build_string, channel };
+            return pkg_info;
+        }
+
+        PackageDiff PackageDiff::get_revision_pkg_diff(
+            std::vector<History::UserRequest> user_requests,
+            int target_revision
```

Comment:
Is `target_revision` allowed to be 0 or negative? Is `user_requests` allowed to be empty?
If there are requirements on these, consider adding assertions (using `assert`) at the beginnig of the function checking that these requirements are fullfilled.